tikv-client/src/main/java/com/pingcap/tikv/region/RegionManager.java

```diff
@@ -184,26 +185,41 @@ public TiRegion getRegionById(long regionId) {
   }
 
   public Pair<TiRegion, Store> getRegionStorePairByKey(ByteString key) {
+    return getRegionStorePairByKey(key, TiStoreType.TiKV);
+  }
+
+  public Pair<TiRegion, Store> getRegionStorePairByKey(ByteString key, TiStoreType storeType) {
     TiRegion region = cache.getRegionByKey(key);
     if (region == null) {
       throw new TiClientInternalException("Region not exist for key:" + formatBytesUTF8(key));
     }
     if (!region.isValid()) {
       throw new TiClientInternalException("Region invalid: " + region.toString());
     }
-    Peer leader = region.getLeader();
-    long storeId = leader.getStoreId();
-    return Pair.create(region, cache.getStoreById(storeId));
-  }
 
-  public Pair<TiRegion, Store> getRegionStorePairByRegionId(long id) {
-    TiRegion region = cache.getRegionById(id);
-    if (!region.isValid()) {
-      throw new TiClientInternalException("Region invalid: " + region.toString());
+    Store store = null;
+    if (storeType == TiStoreType.TiKV) {
+      Peer leader = region.getLeader();
+      store = cache.getStoreById(leader.getStoreId());
+    } else {
+      for (Peer peer : region.getLearnerList()) {
+        Store s = getStoreById(peer.getStoreId());
+        for (Metapb.StoreLabel label : s.getLabelsList()) {
+          if (label.getKey().equals(storeType.getLabelKey())
+              && label.getValue().equals(storeType.getLabelValue())) {
+            store = s;
+            break;
```

Comment:
this `break` only breaks inner loop. Maybe you can use the following code 

```
outerloop:
 for (Peer peer : region.getLearnerList()) {
         Store s = getStoreById(peer.getStoreId());
         for (Metapb.StoreLabel label : s.getLabelsList()) {
           if (label.getKey().equals(storeType.getLabelKey())
               && label.getValue().equals(storeType.getLabelValue())) {
             store = s;
             break outerloop;
            }
         }
   }

```

Reply:
Nice catch. Addressed.